### PR TITLE
Secret masker for agent and command settings.

### PR DESCRIPTION
**Pull request by ericsciple**

### CODE REVIEW

File: src/Agent.Listener/CommandSettings.cs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        {
+            ArgUtil.NotNull(context, nameof(context));
+            _context = context;
+            _promptManager = context.GetService<IPromptManager>();
+            _trace = context.GetTrace(nameof(CommandSettings));
+
+            // Parse the command line args.
+            _parser = new CommandLineParser(
+                hostContext: context,
+                secretArgNames: Constants.Agent.CommandLine.Args.Secrets);
```
Comment (ericsciple) [inline src/Agent.Listener/CommandSettings.cs]: This prevents the command line parser from printing secrets.

Comment (ericsciple) [inline src/Agent.Listener/CommandSettings.cs]: And allows hooking up secret masker to command line parser.

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 2]
```diff
@@ -146,22 +83,15 @@ public AgentSettings LoadSettings()
...
                     // Prompt to acccept the TEE EULA.
-                    acceptTeeEula =
-                        (flags?.Contains(CliArgs.AcceptTeeEula) ?? false) ||
-                        promptManager.ReadBool(
-                            argName: CliArgs.AcceptTeeEula,
-                            description: StringUtil.Loc("AcceptTeeEula"),
-                            defaultValue: false,
-                            args: null,
-                            unattended: unattended);
+                    acceptTeeEula = command.GetAcceptTeeEula();
```
Comment (ericsciple) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: The addition of the CommandSettings class eliminates tedious concern from the configuration manager classes. This allows the configuration manager classes to focus on what they need to do: configure. The tedium of checking supplied flags and prompting details has been moved into the CommandSettings class.

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 3]
```diff
@@ -173,121 +103,67 @@ public AgentSettings LoadSettings()
...
-
-                bool connected = true;
                 try
                 {
                     await TestConnectAsync(serverUrl, creds);
+                    Trace.Info("Connect Complete.");
+                    break;
                 }
-                catch (Exception e)
+                catch (Exception e) when (!command.Unattended)
```
Comment (ericsciple) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: `Unattended` support.

File: src/Agent.Listener/Configuration/ConfigurationManager.cs [hunk 4]
```diff
@@ -173,121 +103,67 @@ public AgentSettings LoadSettings()
...
-                    Trace.Error(e);
-                    _term.WriteLine(StringUtil.Loc("FailedToConnect"));
-                    connected = false;
-                }
-
-                // we don't want to loop when unattended
-                if (unattended || connected)
-                {
-                    break;
+                    _term.WriteError(e);
```
Comment (ericsciple) [inline src/Agent.Listener/Configuration/ConfigurationManager.cs]: I changed this to print the exception message prior to printing our generic "Failed to connect..." messsage.

File: src/Agent.Listener/CommandSettings.cs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            // Check for the arg in the command line parser.
+            ArgUtil.NotNull(validator, nameof(validator));
+            string result;
+            if (!_parser.Args.TryGetValue(name, out result))
+            {
+                result = null;
+            }
+
+            // Return the arg if it is not empty and is valid.
+            _trace.Info($"Arg '{name}': '{result}'");
```
Comment (ericsciple) [inline src/Agent.Listener/CommandSettings.cs]: Tracing added into this class so consumers don't need to sprinkle the same monotonous tracing code all over the place.

File: src/Agent.Listener/Configuration/PromptManager.cs [hunk 6]
```diff
@@ -60,31 +56,13 @@ public override void Initialize(IHostContext hostContext)
             bool secret,
             string defaultValue,
             Func<string, bool> validator,
-            Dictionary<string, string> args,
+            // Dictionary<string, string> args,
```
Comment (ericsciple) [inline src/Agent.Listener/Configuration/PromptManager.cs]: I forgot to remove this line. Will fix.

File: src/Microsoft.VisualStudio.Services.Agent/CommandLineParser.cs [hunk 7]
```diff
@@ -45,28 +50,17 @@ public bool IsCommand(int index, string name)
...
+            _trace.Info(nameof(Parse));
+            ArgUtil.NotNull(args, nameof(args));
+            _trace.Info("Parsing {0} args", args.Length);
 
             string argScope = null;
             foreach (string arg in args)
             {
                 _trace.Info("parsing argument");
-                _trace.Info("arg: {0}", arg);
+                //_trace.Info("arg: {0}", arg);
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/CommandLineParser.cs]: Oops forgot to remove this line.

File: src/Agent.Listener/Agent.cs [hunk 8]
```diff
@@ -95,37 +89,31 @@ public override void Initialize(IHostContext hostContext)
...
 
-                if (parser.Flags.Contains("nostart"))
+                if (command.NoStart)
                 {
-                    Trace.Info("No start option, exiting the agent");
                     return 0;
                 }
 
-                if (parser.IsCommand("run") && !configManager.IsConfigured())
+                if (command.Run && !configManager.IsConfigured())
```
Comment (TingluoHuang) [inline src/Agent.Listener/Agent.cs]: How to reach here?

**Comment by TingluoHuang** [comment]
Look good.

**Pull request merged**